New document/view architecture

slluis opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/595176
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/691254

**REVIEW WARNING: This PR is built on top of new-service-model (PR #7172), a branch that was initially part of this PR but that I moved to a different PR to make it easier to review.**

**Also, it is easier to review this PR commit by commit. Most of the meat is in the two first commits. The other commits are changes to migrate to the new model and to track api changes.**

This PR implements a new architecture for displaying views in the shell. The main changes of this architecture compared to the old one are:

* Introduction of a Model/View/Controller pattern. By having the data in a Model it will be possible to easily support displaying and editing a file in more than one view, even in different documents.
* Extensibility at document and controller level, so that add-ins can plug into file operations and provide custom commands.
* Proper support of split and tabbed views. Extensions can provide visual designers and previews, and the shell decides the best way to present them.

## Document models

A model is the data being shown in a view. DocumentModel is the base class for models that can be edited in a view. FileModel is a base class for file-based models. TextFileModel is the base class for models stored in text files. TextBufferFileModel is a model that stores data in a TextBuffer. DocumentModelRegistry is a service that can be used to share and get shared models.

## Document controllers

The new DocumentController and related classes replace ViewContent and BaseView content. A document controller is a class that implements the logic for loading, displaying and interacting with the contents of a document. The FileDocumentController class is a controller base class specialized in showing the content of files.

DocumentController supports extensions using the DocumentControllerClass. Extensions can hook into controller operations and provide custom UI.

## Document views

Document controllers can provide views using the DocumentView class and derived classes. Unlike the old ViewContent class, which created a Control, the DocumentController class creates a DocumentView. A DocumentView is a component of the shell that can show the content provided by a document controller. The DocumentViewContent subclass can show a Control, and the DocumentViewContainer can show a set of views using tabs or splits.

SdiWorkspaceWindow has changed to support the new DocumentView model. However, all the interaction between DocumentView and the shell implementation is now done through interfaces so that porting to a native toolkit will be easier in the future, and also makes the code easier to unit test.

There are several GtkShellDocumentView* classes that implement the functionality of DocumentView and related classes.

All old ViewContent implementations have been migrated to use DocumentModel and DocumentView.

## Shell documents

Document handling logic has been moved from Workbench to a new DocumentManager class.

The Document class is not a subclass of DocumentContext anymore. It has a DocumentContext property instead. All logic in charge of parsing documents has been moved to the new RoslynDocumentExtension class, a document controller extension that is created only when the document is parseable.

## Unit test infrastructure

The unit test infrastructure has been improved to make it easier to test functionality that depens on the document model. There are new mock classes for IShell and related interfaces. Unit tests now can use the [RequireService] attribute to force the initialization of a service before starting a test.

## TODO

* [ ] More unit tests
   * [ ] Test project reload for views
* [x] Proper UI for view collections that support tabs + split mode
* [x] Translation catalog is not yet initialized when start window is shown
* [ ] Colors of start window are wrong
* [x] Application menu has wrong label

mkrueger committed:
[CSharpBinding] Fixed failing unit test.

slluis committed:
Track API changes

Mostly due to DocumentContext now becoming a property of Document,
and also some missing service reference conversions.

slluis committed:
Moved the text editor to the new document/view architecture

slluis committed:
Implemented Assembly Browser as a document controller

slluis committed:
Implemented build output view as a document controller

slluis committed:
Implemented getting started page as a document controller

slluis committed:
Implemented version control views as document controller extensions

slluis committed:
Fixed F# unit tests

Tests now asynchronously initialize the IDE.

slluis committed:
Implemented hex editor as a document controller

slluis committed:
Implemented gtk designer as a document controller

slluis committed:
Implemented .dot file viewer as document controller

slluis committed:
Implemented .po editor as document controller

slluis committed:
Implemented regex toolkit as document controller

slluis committed:
Implemented disassembly and no-source views as document controllers

slluis committed:
Removed old razor completion code that is not used anymore

slluis committed:
Track API changes

slluis committed:
Migrate tests to the new testing infrastructure

slluis committed:
Implement connected services view as document controller

slluis committed:
Made the method for showing the welcome window asynchronous

slluis committed:
Fix some service initialization issues and track api changes

slluis committed:
Implemented new document/view architecture

Introduced the new DocumentController and related classes, which replaces
ViewContent and BaseView content. A document controller is a class that
implements the logic for loading, displaying and interacting with the contents
of a document. The FileDocumentController class is a controller base class
specialized in showing the content of files.

DocumentController supports extensions using the DocumentControllerClass.
Extensions can hook into controller operations and provide custom UI.

This commit also introduces the DocumentView class and derived classes.
A DocumentView is a component of the shell that can show the content provided
by a document controller. The DocumentViewContent subclass can show a Control,
and the DocumentViewContainer can show a set of views using tabs or splits.

Unlike the old ViewContent class, which created a Control, the
DocumentController class creates a DocumentView.

SdiWorkspaceWindow has changed to support the new DocumentView model.
However, all the interaction between DocumentView and the shell implementation
is now done through interfaces so that porting to a native toolkit will be easier in
the future, and also makes the code easier to unit test.

There are several GtkShellDocumentView* classes that implement the
functionality of DocumentView and related classes.

Document handling logic has been moved from Workbench to a new DocumentManager
class.

The Document class is not a subclass of DocumentContext anymore. It has
a DocumentContext property instead. All logic in charge of parsing documents
has been moved to the new RoslynDocumentExtension class, a document
controller extension that is created only when the document is parseable.

slluis committed:
Unit test infrastructure improvements

Created mock classes for IShell and related interfaces.
Unit tests now can use the [RequireService] attribute to force the
initialization of a service before starting a test.

slluis committed:
Implemented DocumentModel and derived classes

A model is the data being shown in a view. DocumentModel is the
base class for models that can be edited in a view.
FileModel is a base class for file-based models. TextFileModel is
the base class for models stored in text files. TextBufferFileModel
is a model that stores data in a TextBuffer.
DocumentModelRegistry is a service that can be used to share
and get shared models.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.Text/TextFileUtility.cs:
```diff
@@ -285,17 +305,31 @@ static Stream CreateStream (FilePath tmpPath)
+				if (bom != null && bom.Length > 0)
+					stream.Write (bom, 0, bom.Length);
+			}
+			var sw = new StreamWriter (stream, encoding);
```
Might be better to wrap this in a using and set the `leaveOpen: true`

slluis replied:
Fixed.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs:
```diff
@@ -0,0 +1,496 @@
+		{
+			ImmutableList<DocumentModel> linkedModels = ImmutableList<DocumentModel>.Empty;
+			Dictionary <Type, ModelRepresentation> representations = new Dictionary<Type, ModelRepresentation> ();
+			SemaphoreSlim dataLock = new SemaphoreSlim (1);
```
Unless we set the maximum count too, releasing dataLock twice in a row will allow 2 concurrent accesses.

```suggestion
			SemaphoreSlim dataLock = new SemaphoreSlim (1, 1);
```

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs:
```diff
@@ -0,0 +1,496 @@
+
+			public object Id { get; set; }
+
+			public async Task LinkNew (DocumentModel model)
```
Should these take CancellationTokens?

slluis replied:
Not worth it.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs:
```diff
@@ -0,0 +1,496 @@
+				return representation;
+			}
+
+			public async Task<ModelRepresentation> GetRepresentation (Type type)
```
Should these take CancellationTokens?

slluis replied:
I don't think it is worth making this cancellable. It should be fast.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs:
```diff
@@ -0,0 +1,496 @@
+				return representation;
+			}
+
+			public async Task<bool> Unlink (DocumentModel model)
```
Should these take CancellationTokens?

slluis replied:
No, it this operation can't be cancelled.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs:
```diff
@@ -0,0 +1,496 @@
+					// Any operation that requires the lock of several representation must take
+					// the model lock first, to avoid deadlocks
+					if (!dataLocked) {
+						await dataLock.WaitAsync ();
```
should we do these in parallel?

```
await Task.WhenAll (dataLock.WaitAsync(), targetRep.WaitHandle.WaitAsync(), sourceModel.WaitHandle.WaitAsync())`
```

slluis replied:
I don't think so. We need to take the locks always in the same order, to avoid deadlocks.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs:
```diff
@@ -0,0 +1,496 @@
+
+			void RaiseChangedEvent (Type repType)
+			{
+				foreach (var m in linkedModels.Where (m => m.RepresentationType == repType))
```
Can we just roll in the lambda here so we don't capture the local?

slluis replied:
Fixed.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs:
```diff
@@ -0,0 +1,496 @@
+
+			ModelRepresentation lastChangedRepresentation;
+			int changeVersion;
+			object changeLock = new object ();
```
readonly here to ensure lock integrity.

slluis replied:
Fixed.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModelRegistry.cs:
```diff
@@ -0,0 +1,131 @@
+	[DefaultServiceImplementation]
+	public class DocumentModelRegistry: Service
+	{
+		SemaphoreSlim dataLock = new SemaphoreSlim (1);
```
same reasoning as the other semaphore.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/FileModel.cs:
```diff
@@ -0,0 +1,168 @@
+
+			protected override void OnCreateNew ()
+			{
+				content = new byte [0];
```
`Array.Empty<byte> ()`

slluis replied:
Fixed.

Therzok commented on main/tests/Ide.Tests/MonoDevelop.Ide.FindInFiles/SearchResultWidgetTests.cs:
```diff
@@ -29,18 +29,19 @@
 	{
 		[SetUp]
-		public void SetUp ()
+		public async Task SetUp ()
```
Empty method?

slluis replied:
Fixed.

Therzok commented on main/src/addins/MonoDevelop.RegexToolkit/MonoDevelop.RegexToolkit/ValidateActionCodeRefactoringProvider.cs:
```diff
@@ -89,9 +89,9 @@ public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContex
-					ShowRegexToolkitHandler.RunRegexWindow ().Regex = regex.Value.ToString ();
-					return Task.FromResult(document);
+				async t2 => {
+					(await ShowRegexToolkitHandler.RunRegexWindow ()).Regex = regex.Value.ToString ();
```
Should we explicitly move to the UI thread here?

Therzok committed:
Update main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentModel.cs

Co-Authored-By: slluis <llsan@microsoft.com>

Therzok commented on main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl.csproj:
```diff
@@ -516,6 +515,7 @@
     <Compile Include="Gui\MonoDevelop.VersionControl.UrlBasedRepositoryEditor.cs" />
     <Compile Include="Gui\MonoDevelop.VersionControl.Views.DiffWidget.cs" />
     <Compile Include="Gui\MonoDevelop.VersionControl.Views.LogWidget.cs" />
+    <Compile Include="MonoDevelop.VersionControl\VersionControlDocumentController.cs" />
```
This file is not in the PR

slluis replied:
I accidentally added it in the service model PR: https://github.com/mono/monodevelop/pull/7172/commits/3f08768e1c48470d3c77365bf1225e2f48ad8fbc

Therzok commented on main/tests/Ide.Tests/MonoDevelop.Ide.Editor/DocumentReloadTests.cs:
```diff
@@ -86,13 +86,10 @@ public void ShowFileChangedWarning (bool multiple)
-				ContentName = fileName;
-				return Task.FromResult (true);
+				var content = await TextFileUtility.ReadAllTextAsync (FilePath);
+				Document.Editor.Text = (await TextFileUtility.ReadAllTextAsync (FilePath)).Text;
```
```csharp
Document.Editor.Text = content.Text;
```

slluis replied:
Fixed.

slluis committed:
Properly initialize SemaphoreSlim

Unless we set the maximum count too, releasing dataLock twice in a row will
allow 2 concurrent accesses.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+			// Ensure conditions are re-evaluated.
+			extensionContext = CreateExtensionContext ();
+
+			using (extensionChain.BatchModify ()) {
```
:heart:

slluis replied:
:)

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+		/// <param name="type">Type.</param>
+		protected virtual bool OnCanAssignModel (Type type)
+		{
+			return Model?.GetType () == type;
```
Should we use IsAssignableFrom here for people who extend controllers?

slluis replied:
I guess so.

slluis replied:
I'm going to change this to return false by default. If model replacement is supported, the document controller should explicitly say so by overriding OnCanAssignModel.

slluis replied:
Fixed.

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/Document.cs:
```diff
@@ -28,357 +28,268 @@
+			LastTimeActive = DateTime.Now;
+			this.window.CloseRequested += Window_CloseRequested;
+
+			var workspace = Runtime.PeekService<RootWorkspace> ();
```
Should we not do this when the service is initialized?

slluis replied:
Yes, I'll fix.

Therzok replied:
I'm thinking we should also be returning a registration token from `RunWhenServiceInitialized` which we can dispose if we don't have the service initialized by the time the document is disposed.

Therzok replied:
i.e.: https://github.com/mono/monodevelop/pull/7172/files#r257746754

The reasoning for this is that for example, the service callback list will keep the document alive due to the Func capturing this. So we need a way to clear that so the doc actually goes away.

slluis replied:
Heh, I implemented the registration token before seeing your comment. Yes, the reason I used PeekService here was to avoid holding the document reference forever. In practice this will never happen, but using WhenInitialized and canceling it when disposing the document is the right thing to do.

slluis replied:
Fixed by https://github.com/mono/monodevelop/pull/7172/commits/36b90fc92f1649013378b160ec602aab0111250f.

slluis replied:
Well, not yet fixed.

slluis replied:
Fixed.

slluis committed:
Merge branch 'new-service-model' into new-doc-model

slluis committed:
Fix TaskStore initialization issue

slluis committed:
Address PR feedback

slluis committed:
Merge branch 'new-service-model' into new-doc-model

slluis committed:
Fix some ide initialization issues

Removed some extensions that don't exist anymore

slluis committed:
Merge branch 'new-service-model' into new-doc-model

mhutch commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+		/// Owner project or solution item
+		/// </summary>
+		/// <value>The owner.</value>
+		public WorkspaceObject Owner {
```
What about documents with multiple owners e.g. documents that are in multiple projects?

slluis replied:
This represents the currently active owner, which is what really matters. If the document belongs to more than one project, there will be a selector for the active owner.

mhutch commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+		/// Icon of the document tab
+		/// </summary>
+		/// <value>The stock icon identifier.</value>
+		protected string DocumentIconId {
```
Shouldn't this be an `IconId `?

slluis replied:
I want to make it clear that this is the icon that will be shown in the document tab, which may be different from an icon for the view tab (in case of a document with multiple views).

mhutch replied:
I mean the type, instead of `string`

mhutch commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+		/// <summary>
+		/// Icon of the document tab
+		/// </summary>
+		/// <value>The stock icon identifier.</value>
```
Bad copy/paste

mhutch commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+			}
+		}
+
+		public bool ShowNotification { get; internal set; }
```
What does this do?

slluis replied:
That's a TODO, not yet implemented.

mhutch commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+			}
+		}
+
+		void Model_Changed (object sender, EventArgs e)
```
Isn't this redundant?

slluis replied:
Ah yes, that's some left over code.

mhutch commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentController.cs:
```diff
@@ -0,0 +1,917 @@
+		{
+		}
+
+		ExtensionContext CreateExtensionContext ()
```
Shouldn't this context have some conditions - mimetype, file extension, project type, etc?

mhutch replied:
Ah, I see, some filtering attributes are implemented on ExportDocumentControllerBaseAttribute. Would be nice to have project type and project capabilities (and `AppliesToAttribute`).

slluis replied:
Good point! I'll add those.

mhutch commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Documents/DocumentViewContainer.cs:
```diff
@@ -0,0 +1,275 @@
+		/// Gets or sets the display modes supported by this container
+		/// </summary>
+		/// <value>The supported modes.</value>
+		public DocumentViewContainerMode SupportedModes {
```
This entire concept of display modes is very unclear to me. Do you have docs explaining how this works?

mhutch replied:
Specifically, I'd like to know how display modes factor into the following cases:

* Android designer with tabs for source, split, and design
* Xaml previewer with tabs for source and split

slluis replied:
In the case of the designer and the previewer, it would set both tabs and split as supported modes. How those modes will be presented to the user is still TBD. One simple to implement option is to show the tabs, with an extra "Split" tab that shows source and design in a split, which is what the initial PR from David was doing. Another option is to have a button in the bottom bar for switching between tab and split modes. We'll have to discuss this with the UX board.

mhutch replied:
We had a fairly lengthy UX discussion before settling on the approach in David's PR, I'm not sure I want to reopen that can of worms :)

Is there any way to express that only "source" and "source/preview split" views are valid, and that the "preview" view cannot be displayed outside of the split? Or to express what the split view should be labeled?

slluis replied:
No, there is no way to express that "preview" can't be displayed outside of the split. I don't see why add this restriction. Also, there is no way to indicate what the split label should be.

mhutch replied:
Well, in the case of the XAML previewer for example, the preview by itself is not particularly useful. So makes sense to remove that tab, and call the split mode "Preview".

slluis replied:
I still think that even in the XAML previewer case makes sense. But ok, let me think about it. It could be supported with an additional mode.

mhutch replied:
I agree that it _could_ be useful, but that doesn't mean it's useful enough to justify adding additional complexity to the UI in the form of another button/option. As far as I know no-one's _asked_ for this feature, after all :)

slluis committed:
Fix build

slluis committed:
Merge branch 'new-service-model' into new-doc-model

slluis committed:
Fix NRE when initializing workbench

slluis committed:
Include basic unit test infrastructure in the bundle

slluis committed:
Merge branch 'new-service-model' into new-doc-model

slluis committed:
Fix issue when changing the project of a document

Properly propagate the owner change event to extensions.

slluis committed:
Fix NRE in pathed document tests

slluis committed:
Fix NRE in editor

It may happen when the task service is not initialized, when running tests.

slluis committed:
Move more services to IdeServices

And track api changes

slluis committed:
Fix document focusing issue

The active document was not properly set when opening the first document.

slluis committed:
Fix several issues with document extensions

And add unit tests

slluis committed:
Fix document extension issues

Fixed issues with OnContentChanged and OnOwnerChanged. Those methods
are now properly called when changes happen in the controller. Also added
and fixed tests.

Fixed several issues in the GTK designer.

Therzok commented on main/tests/Ide.Tests/MonoDevelop.Ide.Gui.Documents/DocumentControllerTests.cs:
```diff
@@ -0,0 +1,147 @@
+namespace MonoDevelop.Ide.Gui.Documents
+{
+	[TestFixture]
+	public class DocumentControllerTests: IdeTestBase
```
Nice

slluis committed:
Fix main menu

For some reason assigning the main menu before adding the top items
causes a weird application menu item to show up.

slluis committed:
Set correct initial size for split views

slluis committed:
Add UI for view collections that support tabs + split mode

slluis committed:
DocumentController fixes

Fix some issues with HasUnsavedChanges property. Models now have an event
to detected changes in that property. However, that event is not used in the old
editor since it is not reliable due to how undo/redo operations are implemented.

Properly implemented ShowNotification property.

slluis committed:
Remove the DiscardChanges method from Document

It is ambiguous and not really necessary.

slluis committed:
Add new IProjectPadNodeSelector interface

It can be used to provide the node object to be selected in the
solution pad when a document is selected in the shell.

slluis committed:
Merge branch 'new-service-model' into new-doc-model

slluis committed:
Add OnContentShown/Hidden methods to DocumentController

Those methods are invoked when content is shown or hidden in the shell.
Added corresponding infrastructure to DocumentView and derived classes.

slluis committed:
Fix unit tests caused by missing service initialization

slluis commented:
I'll be back!
